compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUUnsupportedScalabilityToLoops.cpp

```diff
@@ -0,0 +1,138 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/Common/TileSizeSelection.h"
+#include "iree/compiler/Codegen/LLVMCPU/PassDetail.h"
+#include "iree/compiler/Codegen/LLVMCPU/Passes.h"
+#include "iree/compiler/Codegen/LLVMCPU/Utils.h"
+#include "iree/compiler/Codegen/Utils/Utils.h"
+#include "mlir/Dialect/SCF/Utils/Utils.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#define DEBUG_TYPE "iree-llvmcpu-unsupported-scalability-to-loops"
+#define VEC_DBGS() (llvm::dbgs() << '[' << DEBUG_TYPE << "] ")
+
+namespace mlir::iree_compiler {
+
+namespace {
+
+class LLVMCPUUnsupportedScalabilityToLoopsPass
+    : public LLVMCPUUnsupportedScalabilityToLoopsBase<
+          LLVMCPUUnsupportedScalabilityToLoopsPass> {
+public:
+  using LLVMCPUUnsupportedScalabilityToLoopsBase::
+      LLVMCPUUnsupportedScalabilityToLoopsBase;
+
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry
+        .insert<arith::ArithDialect, linalg::LinalgDialect, scf::SCFDialect>();
+  }
+
+  void runOnOperation() override;
+};
+
+static bool opKnownToSupport2DScalableVectorizationWithArmSME(Operation *op) {
+  return isa<linalg::MatmulOp, linalg::MatmulTransposeAOp, linalg::FillOp>(op);
+}
+
+struct DropUnsupportedScalableDimsFromTilingInterfaceOps
+    : public OpInterfaceRewritePattern<TilingInterface> {
+  DropUnsupportedScalableDimsFromTilingInterfaceOps(MLIRContext *context,
+                                                    bool assumeArmSME)
+      : OpInterfaceRewritePattern(context), assumeArmSME(assumeArmSME) {}
+
+  LogicalResult matchAndRewrite(TilingInterface op,
+                                PatternRewriter &rewriter) const override {
+    // Note: This rewrite is currently only required for ArmSME (which is the
+    // only target that currently has some concept of 2D scalability).
+    auto targetAttr = IREE::HAL::ExecutableTargetAttr::lookup(op);
+    bool isArmSME = assumeArmSME || hasSMEFeature(targetAttr);
+    if (!isArmSME || opKnownToSupport2DScalableVectorizationWithArmSME(op))
+      return failure();
+
+    auto loweringConfigAttr = getLoweringConfig(op);
+    if (!loweringConfigAttr)
+      return failure();
+
+    auto tileSizes = loweringConfigAttr.getTileSizeVals();
+    auto scalableFlags = loweringConfigAttr.getScalableTileFlagVals();
+
+    // Drop scalable dimensions from leading tiling levels first (and leading
```

Comment:
```suggestion
    // 1. Drop scalable dimensions from leading tiling levels first (and leading
```

This is a bit complex, hence suggesting to clearly mark separate steps.